src/cargo/ops/cargo_compile.rs

```diff
@@ -23,26 +23,24 @@
 //!
 
 use std::collections::HashSet;
-use std::path::{Path, PathBuf};
+use std::path::PathBuf;
 use std::sync::Arc;
 
 use core::{Package, Source, Target};
 use core::{PackageId, PackageIdSpec, Profile, Profiles, TargetKind, Workspace};
-use core::compiler::{BuildConfig, Compilation, Context, DefaultExecutor, Executor};
+use core::compiler::{BuildConfig, Compilation, CompileMode, Context, DefaultExecutor, Executor};
 use core::compiler::{Kind, Unit};
 use core::resolver::{Method, Resolve};
 use ops;
 use util::config::Config;
-use util::{profile, CargoResult, CargoResultExt};
+use util::{profile, CargoResult};
 
 /// Contains information about how a package should be compiled.
 #[derive(Debug)]
 pub struct CompileOptions<'a> {
     pub config: &'a Config,
-    /// Number of concurrent jobs to use.
-    pub jobs: Option<u32>,
-    /// The target platform to compile for (example: `i686-unknown-linux-gnu`).
-    pub target: Option<String>,
+    /// Configuration information for a rustc build
+    pub build_config: BuildConfig,
```

Comment:
So, at https://github.com/rust-lang/cargo/pull/5351#discussion_r181147836 @alexcrichton suggested that we don't do too much stuff when parsing command line arguments. I am not sure I entirely agree, to me it seems beneficial to lower stringly-typed command-line flags into specific data structures as early as possible, even if that requires looking at a broader context then just cli. However in this case creating a `BuildConfig` requires running `rustc`, if only to get `host`, which seems a bit too heavy weight for command-line parsing. 

Perhaps we could use an API like `CompileOptions::into_build_config(self) -> BuildConfig`, and call that when constructing a context?

Reply:
The way I understand @alexcrichton's remarks is that we shouldn't do any input validation beyond upcasting to more specific types, because in many cases we want API users to benefit from the same validation that the CLI needs. So, where we can employ types to do validation, we should (this makes it ~impossible for API users to pass invalid input anyway), but validation of actually string-typed data should be done on the inside of the library API.

I understand and agree with the reluctance to run `rustc` during the construction of `CompileOptions`, but I'm not sure I see much value in having a `CompileOptions::into_build_config(self)` API -- this would still duplicate a bunch of stuff in `CompileOptions` (which gets deconstructed in `cargo::ops::compile()`) and the `BuildConfig` itself.

It still feels to me like the `Rustc` does not belong in `BuildConfig`; previously, `BuildConfig` was a straightforward bag of data about build configuration, but `Rustc` is not like that -- it's more a source of build configuration instead. I would be more inclined to move the `Rustc` and things that depend on it (the target configs `host` and `target`) into the `Context`. This has the added advantage that the construction of the `TargetInfo`s gets closer to these other things, so we could eventually merge those calls to `rustc`. If we do this, `BuildConfig` can be (cheaply) constructed as part of `CompileOptions`. What do you think?

Reply:
Hm, yeah, moving `rustc`, `host` and `target` out of `BuildConfig` seems reasonable. On the other after #5384 the build config would look like

```
/// Configuration information for a rustc build.
pub struct BuildConfig {
    pub rustc: Rustc,
    /// Build information for the host arch
    pub host: TargetConfig,
    /// The target arch triple, defaults to host arch
    pub requested_target: Option<String>,
    /// Build information for the target
    pub target: TargetConfig,
    /// How many rustc jobs to run in parallel
    pub jobs: u32,
    /// Whether we are building for release
    pub release: bool,
    /// Whether to print std output in json format (for machine reading)
    pub json_messages: bool,
}
```

and, if we move host/target out of it, it becomes really thin. Why do we need in the first place? Could we just fold it all into context?

Reply:
That would make `Context` even more of a God object. I've been looking at angles to splitting up `Context`, maybe in a part that's more stateless (does not depend on the units submitted) and a part that's more stateful. With #5384 pending, I'll probably wait until that settles, but in the meantime I suggest we avoid moving more stuff there unless it improves the design. Worst case, `BuildConfig` is still a useful grouping of things that live together in `CompileOptions`.

Reply:
>I still feel there are a number of improvements in this PR that are worth landing, even if they make BuildConfig a bit thin for now.

Just to make clear, you'd like lift `rustc` from `BuildConfig` to `Context`, so that it is not created during cli parsing, and otherwise leave the rest as it is now? Sounds good to me, I think we can land this before #5384 than.

cc @ehuss

Reply:
#5384 came to fruition faster that this PR, so merging that!